pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkSVForwardIndexWriterV4.java

```diff
@@ -0,0 +1,235 @@
+    } finally {
+      CleanerUtil.cleanQuietly(buffer);
+    }
+  }
+
+  private void flushChunk() {
+    if (_nextDocId > _docIdOffset) {
+      writeChunk();
+    }
+  }
+
+  private void writeChunk() {
+    /*
+    This method translates from the current state of the buffer, assuming there are 3 values of lengths a,b, and c:
+    [-][a][a bytes][b][b bytes][c][c bytes]
+    to:
+    [3][16][a+16][a+b+16][a bytes][b bytes][c bytes]
+    [------16-bytes-----][----a+b+c bytes----------]
+     */
+    int numDocs = _nextDocId - _docIdOffset;
+    _chunkBuffer.putInt(0, numDocs);
+    // collect lengths
+    int[] valueLengths = new int[numDocs];
+    int[] offsets = new int[numDocs];
+    int offset = Integer.BYTES;
+    for (int i = 0; i < numDocs; i++) {
+      offsets[i] = offset;
+      int size = _chunkBuffer.getInt(offset);
+      valueLengths[i] = size;
+      offset += size + Integer.BYTES;
+    }
+    // now iterate backwards shifting variable length content backwards to make space for prefixes at the start
+    // this pays for itself by allowing random access to readers
+    int limit = _chunkBuffer.position();
+    int accumulatedOffset = Integer.BYTES;
+    for (int i = numDocs - 2; i >= 0; i--) {
+      ByteBuffer source = _chunkBuffer.duplicate();
+      int copyFrom = offsets[i] + Integer.BYTES;
+      source.position(offsets[i]).limit(copyFrom + valueLengths[i]);
+      _chunkBuffer.position(offsets[i] + accumulatedOffset);
+      _chunkBuffer.put(source.slice());
+      accumulatedOffset += Integer.BYTES;
+    }
+    // compute byte offsets of each string from lengths
+    int metadataOffset = Integer.BYTES * (numDocs + 1);
+    offsets[0] = metadataOffset;
+    int cumulativeLength = valueLengths[0];
+    for (int i = 1; i < offsets.length; i++) {
+      offsets[i] = metadataOffset + cumulativeLength;
+      cumulativeLength += valueLengths[i];
+    }
+    // write the lengths into the space created at the front
+    for (int i = 0; i < offsets.length; i++) {
+      _chunkBuffer.putInt(Integer.BYTES * (i + 1), offsets[i]);
+    }
+    _chunkBuffer.position(0);
+    _chunkBuffer.limit(limit);
+    write(_chunkBuffer, false);
+    clearChunkBuffer();
+  }
+
+  private void write(ByteBuffer buffer, boolean huge) {
+    int maxCompressedSize = _chunkCompressor.maxCompressedSize(buffer.limit());
+    ByteBuffer target = null;
+    try {
+      target = _dataChannel.map(FileChannel.MapMode.READ_WRITE, _chunkOffset, maxCompressedSize)
+          .order(ByteOrder.LITTLE_ENDIAN);
+      int compressedSize = _chunkCompressor.compress(buffer, target);
+      // reverse bytes here because the file writes BE and we want to read the metadata LE
+      _output.writeInt(Integer.reverseBytes(_docIdOffset | (huge ? 0x80000000 : 0)));
+      _output.writeInt(Integer.reverseBytes((int) (_chunkOffset & 0xFFFFFFFFL)));
+      _metadataSize += 8;
+      _chunkOffset += compressedSize;
+      _docIdOffset = _nextDocId;
+    } catch (IOException e) {
```

Comment:
For a write method, might be better to throw the `IOException` and expect the caller to handle it

Reply:
See above.